### ISSUE TITLE
Importing .sav from NDS doesn't work

### ISSUE THREAD

**Issue by Felina-Lain** [Arisotura/melonDS#777]
Hello

I'm trying to import some old games with .sav from my nds
Importing them in Desmume works (and when it doesn't I can change the file size so it does)

But for a lot of them, importing into Melon either does nothing (melon doesn't seem to see the file at all) or it give me an error message:
"The save data could not be accessed. Please turn the power off and reinsert the SD card"

I think it might be that the file size detection doesn't work right
(Though I seen in FAQ and forum for MelonDS, that if you import from DS .sav the software will use that to determine file size? But apparently it doesn't work right)

I'm not quite sure though (if there is a way to give you logs, just tell me how to get them, and I'll post them here to help)

thanks in advance anyway, because in terms of speed Melon is now leagues above Desmume

**Comment by RSDuck** [comment]
Do the savefiles originate from desmume or from a flashcart?

> The save data could not be accessed. Please turn the power off and reinsert the SD card

are you trying to start them via DSi mode?

**Comment by Felina-Lain** [comment]
They're from a flashcart, all straight from my NDS where they worked last I played those games

I'm using normal NDS emulation, not DSI

Weird thing is: some of the games that Desmume could NOT read the save for, Melon does it just fine
But some of the game that Desmume COULD read the save for (although with me specifying the size of the file), Melon can't read them right

I think there was somewhere a thread about using a hex editor to correct the files or something? Would that work?

**Comment by RSDuck** [comment]
flashcarts tend to pad their savefiles to a common size. You can either trim the size of the savefile down using a hexeditor to the size of one created by melonDS or you can try downloading build from [the latest source](https://github.com/Arisotura/melonDS/runs/1194845122)(under artefacts) where an import savefile function is implemented. Though make sure you don't have a savefile with the wrong size, so just let melonDS create one.

> I'm using normal NDS emulation, not DSI

I was wondering because you wrote "Please turn the power off and reinsert the SD card"

**Comment by Felina-Lain** [comment]
> you can try downloading build from [the latest source](https://github.com/Arisotura/melonDS/runs/1194845122)(under artefacts) where an import savefile function is implemented. Though make sure you don't have a savefile with the wrong size, so just let melonDS create one.

I downloaded this version and set it up in a new folder, I set the game in a separate folder too, and started it.
Melon created a savefile. Then I used import savefile
It failed and gave me the message
"Savefile shorter than SRAM"



I'm going to try the hex editor later this week-end, I'll keep you updated

**Comment by RSDuck** [comment]
oh that could mean a few things:
- the entry in melonDS's save type database is wrong. Could you name me the game?
- you still have savefile which is too large lying around so melonDS used this size as a reference
- the savefile incomplete

**Comment by Felina-Lain** [comment]
> * the entry in melonDS's save type database is wrong. Could you name me the game?

I got a dozen games with that problem apparently, but the one I used to test the import function was
Trauma Center: Under the Knife 2

> * you still have savefile which is too large lying around so melonDS used this size as a reference

Lying around, would it matter even in a completely separate folder? I thought it might, which is why I separated all my test files into a completely new folder, so I don't think that would be it?

> * the savefile is incomplete
But it still work on flashcart? Or on Desmume? So I don't think that's it.

I'm going to try download a hex editor, I'll let Melon create a new file, at the size it want, then I'll use that as a reference to trim my own file the right way and try load it. I'll tell you if it works.

**Comment by RSDuck** [comment]
> Lying around, would it matter even in a completely separate folder? I thought it might, which is why I separated all my test files into a completely new folder, so I don't think that would be it?

I meant that it's the savefile melonDS will load once you start the game (so they're in the same folder one and for a rom named romname.nds, the savefile is named romname.sav).

**Comment by Felina-Lain** [comment]
> I meant that it's the savefile melonDS will load once you start the game (so they're in the same folder one and for a rom named romname.nds, the savefile is named romname.sav).

Ah, so no, that can't be it either, since I had the game in a new empty folder, and the savefile I was trying to import was on an entirely separate drive

I'm trying the hex method right now, we'll see how it works

**Comment by RSDuck** [comment]
well that's weird. Import savefile should do the exact same thing as trimming the file manually down. Can you give me some sample files where import savefile didn't work?

**Comment by Felina-Lain** [comment]
> well that's weird. Import savefile should do the exact same thing as trimming the file manually down. Can you give me some sample files where import savefile didn't work?

Well I didn't test them all before just using the hex editor no matter what, but one that I did test for sure was Trauma Center 2
Here is my savefile for it, straight from the flashcart:
https://we.tl/t-j3EZRKJGlp
If I import it, it give me the "Shorter than SRAM" error
If I trim it myself, then import it, it says "Trimmed savefile" and it work, I get my old saves, I can keep playing etc

**Comment by RSDuck** [comment]
hm that's weird, the import savefile function works fine for me (though I noticed one bug namely that unless you save in game after importing a savefile the changes don't end up in the savefile). The savefile melonDS creates for me for Trauma Center 2 (US version) is 8 kb in size.

**Comment by Felina-Lain** [comment]
> hm that's weird, the import savefile function works fine for me (though I noticed one bug namely that unless you save in game after importing a savefile the changes don't end up in the savefile). The savefile melonDS creates for me for Trauma Center 2 (US version) is 8 kb in size.

Well dunno why it didn't for me with flashcart file, but I got it to import in the end with the hex editor, and I'm now enjoying all my old games, so I'm not going to ask for more XD

Also, yeah, I mentioned that in one of the above reply. A good solution I found is to make a change in the options? Like change the sound lower or higher. Since most games save when you do that. 

Anyway, thanks for taking the time to help! and thanks again for the emulator!
